reflection.md

```diff
@@ -15,4 +18,40 @@ Write a few brief paragraphs that discuss:
+
+On the other hand, it seems that it is fairly difficult to write regular expressions when there are patterns that aren't easily represented using repetition. For example, the code we were given had the following example:
+
+```
+val digit = '0' || '1' || '2' || '3' || '4' || '5' || '6' || '7' || '8' || '9' 
+```
+
+This code made me a bit sad because there is an obvious pattern, but we were forced to union 10 literals. In general, union and concatenation operations can create clunky regular expressions. Not that they by themselves are bothersome, but when things get cumbersome, I believe these two operations are deeply involved.
+
+Another trait about our DSL that slightly bothers me is how we have to explicitly use parentheses to correctly express our regular expression:
+
+```
+val pattern = "42" || ( ('a' <*>) ~ ('b' <+>) ~ ('c'{3}))
+```
+
+In this code, we are forced to put three pairs of parentheses to write this regular expression. It would be interesting to see if we can create some sort of rules that allows us to remove parentheses. Some parentheses can be good for clarifying the regular expressions, but for those that seem redundant, removing parentheses could be nice.
+
+Besides removing parentheses, there are three other things I would consider modifying. One is to further simplify the syntax of the DSL. For example, it would be nice to write just a `*` in front of a regular expression like:
+
+```
+("42" || "6")*
+```
+to say I want 0 or more copies of what is inside the parentheses. However, I would like to mention that this change is probably not possible in Scala. It would probably be fairly difficult or impossible) to get Scala to recognize the `*` operator in the code above.
+
+The second thing is to implement some sort of pattern generation within regular expressions. From the example above that takes the union of 10 literals 9 times, it might be nice to have something like a generator or a declarative approach that creates the values we want like in the following line:
+
+```
+val = (Integers from 1 to 8)
+```
+
+where the value inside the parentheses is supposed to represent any digits from 1 to 8.
+
+The last thing I was thinking about was having support to not include certain expressions in a regular expression. For example, it would be nice to able modify `digit` so that it contains all digits 1-9 not including 3, 4, and 7. This was just a side thought however, so I don't want to elaborate on this idea too much. I'd hypothesize, however, that it might be difficult to implement into the DSL.
```

Comment:
You probably want some concept of a range or pattern. For example, `0-9` would mean "the union of all of the characters from 0 to 9, inclusive". Many languages also have character classes, i.e. the pattern `"[0-9a-fA-F\.#]"` represents any digit, or any character from a to f (upper and lower case), or a period, or a pound sign.